tutorials/03_moduly.md

```diff
@@ -0,0 +1,356 @@
+moduly
+======
+
+Zatím jsme tvořili programy v Pythonu tak nějak na divoko, tedy v jednom nebo
+více souborech bez nějakého zvláštního řádu. Na dnešním cvičení se podíváme na
+to, jak tvořit redistribuovatelné moduly, které jdou instalovat pomocí pipu.
+
+Za příklad si vezmeme kód Ondřeje Caletky, který umožňuje určit české svátky
+v zadaném roce. Jako příklad je ideální, protože obsahuje jak funkce, které
+můžeme volat z Pythonu, tak lze volat z příkazové řádky.
+
+ * [oskar456/isholiday.py](https://gist.github.com/oskar456/e91ef3ff77476b0dbc4ac19875d0555e)
+
+setup.py
+--------
+
+Základním stavebním kamenem Python balíčku je soubor `setup.py`, který
+obsahuje všechna potřebná metadata ve volání funkce `setup()` z modulu
+`setuptools`.
+
+Pojďme vytvořit jeho minimální variantu:
+
+```python
+from setuptools import setup
+
+
+setup(
+    name='isholiday',
+    version='0.1',
+    description='Finds Czech holiday for given year',
+    author='Ondřej Caletka',
+    author_email='ondrej@caletka.cz',
+    license='Public Domain',
+    url='https://gist.github.com/oskar456/e91ef3ff77476b0dbc4ac19875d0555e',
+    py_modules=['isholiday'],
+)
+```
+
+Všimněte si, že jsme balíček pojmenoval stejně jako soubor se zdrojovým kódem.
+Je to dobrá konvence, ale není to technicky nutné.
+
+Balíček můžeme zkusit nainstalovat do virtualenvu:
+
+```bash
+$ python3.5 -m venv env
+$ . env/bin/activate
+(env)$ python setup.py install
+...
+(env)$ python
+>>> import isholiday
+>>> 
+(env)$ python -m pip freeze
+isholiday==0.1
+```
+
+Přes `setup.py` můžeme dělat další věci, například vytvořit archiv s balíčkem:
+
+```bash
+(env)$ python3 setup.py sdist
+...
+warning: sdist: standard file not found: should have one of README, README.rst, README.txt
+...
+```
+
+Extra soubory do zdrojového balíčku
+-----------------------------------
+
+Jak vidíte, `setuptools` si stěžuje, že náš projekt nemá `README`.
+Můžeme jej vytvořit a uložit jako `README` přímo v kořenovém adresáři projektu,
+tedy tam, kde byste jej nejspíš čekali.
+
+```
+Czech public holiday checker...
+```
+
+```bash
+(env)$ python3 setup.py sdist
+```
+
+V adresáři `dist` najdete archiv, jeho obsah můžete zkontrolovat, měl by tam
```

Comment:
Nicméně @encukou má rech, určitě bych zmínil, že je to jen obyčejný archiv s jinou příponou, žádný „magic.“